<div style="background-color:#FAF3F3; padding: 20px; border-radius: 10px; box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1); border:2px solid #E57373; margin-top: 20px;">
    <h1 style="font-size:28px; font-family:Georgia, serif; color:#D32F2F; text-align: center; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);">
        Exploratory Data Analysis on  Daily Expense Tracker
    </h1>
</div>

<div style="background-color:#E6F9E6; padding: 20px; border-radius: 10px; box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1); border:2px solid #66C2A5; margin-top: 20px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#4A90E2; text-align: center; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);">
        About Author
    </h1>
    <div style="text-align:center;">
        <a href="https://misbah.devsil.com/">
            <img src="https://misbah.devsil.com/assets/images/picture%20misbah.png" style="width: 20%; border-radius: 50%;" alt="Misbah Khattak">
        </a>
    </div>
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333; text-align:center;">Mr. Misbah Khattak</h2>
    <p style="font-size:16px; font-family:Georgia, serif; line-height: 1.5em; text-align:center; color:#333;">
        <em>Data Scientist / AI Engineer</em><br>
        BS Computer Science
    </p>
    <div style="text-align:center; margin-top: 10px;">
        <a href="https://www.linkedin.com/mynetwork/grow/">
            <img width="50" height="50" src="https://img.icons8.com/color/48/linkedin.png" alt="linkedin">
        </a>
        <a href="https://github.com/misbahkhattak718-pixel?tab=repositories">
            <img width="50" height="50" src="https://img.icons8.com/ios-glyphs/50/github.png" alt="github">
        </a>
        <a href="https://www.kaggle.com/misbahkhattak123">
            <img width="50" height="50" src="https://img.icons8.com/clouds/100/kaggle.png" alt="kaggle">
        </a>
    </div>
    <div style="text-align:center; margin-top: 20px;">
        <a href="https://misbah.devsil.com/" style="font-size:18px; font-family:Georgia, serif; color:#4A90E2; text-decoration:none; border:2px solid #4A90E2; padding: 5px 15px; border-radius: 5px;">
            Portfolio Site
        </a>
    </div>
</div>


<div style="padding: 15px; border-radius: 10px; box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1); border:2px solid #9E9E9E; margin-top: 15px;">
    <h2 style="font-size:22px; font-family:Georgia, serif; color:#333333; text-align: center; text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.1);">
        Table of Contents
    </h2>
    <div style="background-color:#E1F5FE; padding: 10px; border-radius: 5px; margin: 10px 0;">
        <strong><a href="#1" style="text-decoration: none; color: #0D47A1;">1. Introduction</a></strong>
        <div style="font-size:14px; color:#333;">
            <ul>
                <li><a href="#1.1">1.1 Project Description</a></li>
                <li><a href="#1.2">1.2 Data Description</a></li>
                <li><a href="#1.3">1.3 Objectives</a></li>
            </ul>
        </div>
    </div>
    <div style="background-color:#E8F5E9; padding: 10px; border-radius: 5px; margin: 10px 0;">
        <strong><a href="#2" style="text-decoration: none; color: #1B5E20;">2. EDA Steps</a></strong>
        <div style="font-size:14px; color:#333;">
            <ul>
                <li><a href="#2.1">2.1 Overview of Dataset</a></li>
                <li><a href="#2.2">2.2 Basic Statistics</a></li>
                <li><a href="#2.3">2.3 Data Visualization</a></li>
                <li><a href="#2.4">2.4 Correlation Analysis</a></li>
                <li><a href="#2.5">2.5 Income vs. Expense Analysis</a></li>
                <li><a href="#2.6">2.6 Category Analysis</a></li>
                <li><a href="#2.7">2.7 Time Series Analysis</a></li>
                <li><a href="#2.8">2.8 Outlier Detection</a></li>
                <li><a href="#2.9">2.9 Visualizations for Each Column</a></li>
                <li><a href="#2.10">2.10 Insights and Recommendations</a></li>
            </ul>
        </div>
    </div>
    <div style="background-color:#FFF3E0; padding: 10px; border-radius: 5px; margin: 10px 0;">
        <strong><a href="#3" style="text-decoration: none; color: #BF360C;">3. Conclusion</a></strong>
        <div style="font-size:14px; color:#333;">
            <ul>
                <li><a href="#3.1">3.1 Key Findings</a></li>
                <li><a href="#3.2">3.2 Suggestions for Improvement</a></li>
                <li><a href="#3.3">3.3 Next Steps</a></li>
            </ul>
        </div>
    <


<div style="background-color:#D0E6F3; padding: 15px; border-radius: 10px; box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1); border:2px solid #66C2A5; margin-top: 15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align: center; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);">
        1. Introduction
    </h1>
</div>



---
### Project Description

#### Assumptions
1. **Data Integrity**: The dataset contains accurate and reliable financial transaction records. All entries are assumed to be valid and free of missing values.
2. **User Behavior**: The recorded transactions reflect genuine user behaviors, including income and expense patterns.
3. **Time Frame**: The analysis is constrained to the provided dates, primarily focusing on the first quarter of 2022.

#### Methodology
1. **Data Collection**: The project utilizes transaction records from a financial account, detailing various categories such as food, transportation, and household expenses.
2. **Data Cleaning**: The initial step involved cleaning the dataset to handle inconsistencies, including removing duplicates and fixing any anomalies.
3. **Descriptive Statistics**: Statistical measures (mean, median, mode) were calculated to summarize the financial behaviors.
4. **Data Visualization**: Visual tools (bar charts, pie charts, histograms) were employed to illustrate spending trends and patterns.
5. **Correlation Analysis**: Relationships between different categories of expenses and income were explored to identify trends.

#### Scope
- **Analysis Duration**: The analysis will cover the transactions recorded from December 2021 to March 2022. 
- **Focus Areas**: Emphasis will be placed on understanding spending patterns, categorizing expenses, and identifying outliers in the data. 
- **Insights Generation**: The aim is to derive actionable insights that can help in budget planning and financial decision-making.

---
### Data Description



| **Attribute**      | **Description**                                                              | **Example**                        |
|---------------------|------------------------------------------------------------------------------|------------------------------------|
| **Date**            | The date and time when the transaction occurred.                             | 3/2/2022 10:11                     |
| **Account**         | The payment method or account used for the transaction.                     | CUB - online payment               |
| **Category**        | The broad classification of the expense or income.                          | Food                               |
| **Subcategory**     | A more specific classification that provides additional details.            | Snacks                             |
| **Note**            | Additional comments or details about the transaction.                       | Brownie                            |
| **Amount**          | The monetary value of the transaction.                                      | 50                                 |
| **Currency**        | The currency in which the transaction is recorded.                          | INR (Indian Rupees)               |
| **Income/Expense**  | Indicates whether the transaction is classified as income or an expense.   | Expense                            |



### Objectives
1. **Identify Spending Patterns**: Analyze trends in expenditures across different categories and subcategories to understand user behavior.

2. **Visualize Financial Trends**: Create visual representations of income and expenses to track financial health over time.

3. **Highlight Key Performers**: Identify which categories contribute most significantly to the overall spending and income.

4. **Provide Recommendations**: Based on the analysis, suggest potential areas for cost-saving adjustments or increased investment opportunities.

---

<div style="background-color:#E6F2FF; padding: 10px; border-radius: 8px; box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1); border:1px solid #66C2A5; margin-top: 10px; margin-left: 20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align: left; margin: 0;">
        - Importing Libraries
    </h2>
</div>

In [2]:
# import the libreries:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import plotly.express as px

<div style="background-color:#E6F2FF; padding: 10px; border-radius: 8px; box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1); border:1px solid #66C2A5; margin-top: 10px; margin-left: 20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align: left; margin: 0;">
       - Loading Dataset
    </h2>
</div>

In [3]:
# load the dataset:
df = pd.read_csv('../../Practice/01_python_first/expense_data_1.csv')


<div style="background-color:#D0E6F3; padding: 15px; border-radius: 10px; box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1); border:2px solid #66C2A5; margin-top: 15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align: center; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2);">
    2.EDA Steps
    </h1>
</div>

<div style="background-color:#E6F2FF; padding: 10px; border-radius: 8px; box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1); border:1px solid #66C2A5; margin-top: 10px; margin-left: 20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align: left; margin: 0;">
    2.1 Overview of Dataset
    </h2>
</div>

<div style="background-color:#E6F2FF; padding: 10px; border-radius: 8px; box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1); border:1px solid #66C2A5; margin-top: 10px; margin-left: 20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align: left; margin: 0;">
    - Show the first few rows of the datafram:
    </h2>
</div>

In [4]:
df.head()

,Date,Account,Category,Subcategory,Note,INR,Income/Expense,Note.1,Amount,Currency,Account.1
0,3/2/2022 10:11,CUB - online payment,Food,NaN,Brownie,50.0,Expense,NaN,50.0,INR,50.0
1,3/2/2022 10:11,CUB - online payment,Other,NaN,To lended people,300.0,Expense,NaN,300.0,INR,300.0
2,3/1/2022 19:50,CUB - online payment,Food,NaN,Dinner,78.0,Expense,NaN,78.0,INR,78.0
3,3/1/2022 18:56,CUB - online payment,Transportation,NaN,Metro,30.0,Expense,NaN,30.0,INR,30.0
4,3/1/2022 18:22,CUB - online payment,Food,NaN,Snacks,67.0,Expense,NaN,67.0,INR,67.0


<div style="background-color:#E6F2FF; padding: 10px; border-radius: 8px; box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1); border:1px solid #66C2A5; margin-top: 10px; margin-left: 20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align: left; margin: 0;">
    - Show the last few rows of the datafram:
    </h2>
</div>

In [5]:
df.tail()

,Date,Account,Category,Subcategory,Note,INR,Income/Expense,Note.1,Amount,Currency,Account.1
272,11/22/2021 14:16,CUB - online payment,Food,NaN,Dinner,90.0,Expense,NaN,90.0,INR,90.0
273,11/22/2021 14:16,CUB - online payment,Food,NaN,Lunch with company,97.0,Expense,NaN,97.0,INR,97.0
274,11/21/2021 17:07,CUB - online payment,Transportation,NaN,Rapido,130.0,Expense,NaN,130.0,INR,130.0
275,11/21/2021 15:50,CUB - online payment,Food,NaN,Lunch,875.0,Expense,NaN,875.0,INR,875.0
276,11/21/2021 13:30,CUB - online payment,Other,NaN,Got from gobi,2000.0,Income,NaN,2000.0,INR,2000.0


In [6]:
# print the shape of the dataset:
df.shape

(277, 11)

In [9]:
# print the count of columns and rows of the dataset:
print (f"Number of rows: {df.shape[0]}")
print (f"Number of columns: {df.shape[1]}")

Number of rows: 277
Number of columns: 11


In [ ]:
# show the column names:
df.columns

Index(['Date', 'Account', 'Category', 'Subcategory', 'Note', 'INR',
       'Income/Expense', 'Note.1', 'Amount', 'Currency', 'Account.1'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            277 non-null    object 
 1   Account         277 non-null    object 
 2   Category        277 non-null    object 
 3   Subcategory     0 non-null      float64
 4   Note            273 non-null    object 
 5   INR             277 non-null    float64
 6   Income/Expense  277 non-null    object 
 7   Note.1          0 non-null      float64
 8   Amount          277 non-null    float64
 9   Currency        277 non-null    object 
 10  Account.1       277 non-null    float64
dtypes: float64(5), object(6)
memory usage: 23.9+ KB


In [14]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Subcategory,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
INR,277.0,410.750903,1065.756569,3.0,50.0,128.0,301.15,10000.0
Note.1,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Amount,277.0,406.759134,1065.158318,3.0,50.0,125.0,300.00,10000.0
Account.1,277.0,406.759134,1065.158318,3.0,50.0,125.0,300.00,10000.0


In [15]:
# summery statistics of the dataset:
from skimpy import skim

skim(df)

╭──────────────────────────────────────────────── skimpy summary ─────────────────────────────────────────────────╮
│          Data Summary                Data Types                                                                 │
│ ┏━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓ ┏━━━━━━━━━━━━━┳━━━━━━━┓                                                          │
│ ┃ Dataframe         ┃ Values ┃ ┃ Column Type ┃ Count ┃                                                          │
│ ┡━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩ ┡━━━━━━━━━━━━━╇━━━━━━━┩                                                          │
│ │ Number of rows    │ 277    │ │ string      │ 6     │                                                          │
│ │ Number of columns │ 11     │ │ float64     │ 5     │                                                          │
│ └───────────────────┴────────┘ └─────────────┴───────┘                                                          │
│                                                    All null                                                     │
│ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┓  │
│ ┃ column                                                   ┃ NA                   ┃ NA %                     ┃  │
│ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━┩  │
│ │ Subcategory                                              │                  277 │                      100 │  │
│ │ Note.1                                                   │                  277 │                      100 │  │
│ └──────────────────────────────────────────────────────────┴──────────────────────┴──────────────────────────┘  │
│                                                     number                                                      │
│ ┏━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┓  │
│ ┃ column         ┃ NA   ┃ NA %    ┃ mean     ┃ sd      ┃ p0   ┃ p25   ┃ p50   ┃ p75     ┃ p100    ┃ hist     ┃  │
│ ┡━━━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━┩  │
│ │ INR            │    0 │       0 │    410.8 │    1066 │    3 │    50 │   128 │   301.1 │   10000 │    ▇     │  │
│ │ Amount         │    0 │       0 │    406.8 │    1065 │    3 │    50 │   125 │     300 │   10000 │    ▇     │  │
│ │ Account.1      │    0 │       0 │    406.8 │    1065 │    3 │    50 │   125 │     300 │   10000 │    ▇     │  │
│ └────────────────┴──────┴─────────┴──────────┴─────────┴──────┴───────┴───────┴─────────┴─────────┴──────────┘  │
│                                                     string                                                      │
│ ┏━━━━━━━━━━━┳━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┓  │
│ ┃           ┃    ┃           ┃          ┃           ┃          ┃           ┃ chars    ┃ words per ┃ total    ┃  │
│ ┃ column    ┃ NA ┃ NA %      ┃ shortest ┃ longest   ┃ min      ┃ max       ┃ per row  ┃ row       ┃ words    ┃  │
│ ┡━━━━━━━━━━━╇━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━┩  │
│ │ Date      │  0 │         0 │ 2/8/2022 │ 12/30/202 │ 1/1/2022 │ 3/2/2022  │     15.1 │         2 │      554 │  │
│ │           │    │           │ 9:18     │ 1 14:10   │ 21:45    │ 10:11     │          │           │          │  │
│ │ Account   │  0 │         0 │ Cash     │ CUB -     │ CUB -    │ Cash      │     19.9 │         4 │     1105 │  │
│ │           │    │           │          │ online    │ online   │           │          │           │          │  │
│ │           │    │           │          │ payment   │ payment  │           │          │           │          │  │
│ │ Category  │  0 │         0 │ Food     │ Self-deve │ Allowanc │ Transport │     5.85 │         1 │      283 │  │
│ │           │    │           │          │ lopment   │ 

In [17]:
# show the datatypes of the columns:
# print name of columns with their datatypes:
for col in df.columns:
    print (f"{col}:     {df[col].dtype}")

Date:     object
Account:     object
Category:     object
Subcategory:     float64
Note:     object
INR:     float64
Income/Expense:     object
Note.1:     float64
Amount:     float64
Currency:     object
Account.1:     float64


In [19]:
# print the number of unique values of the dataset:
for col in df.columns:
    print (f"{col}:     {df[col].nunique()}")

Date:     240
Account:     2
Category:     13
Subcategory:     0
Note:     159
INR:     138
Income/Expense:     2
Note.1:     0
Amount:     137
Currency:     2
Account.1:     137


# OUTPUT
- The number of rows is 277, and the number of columns is 11.
- The columns are:
 1. Date
 2. Account
 3. Category
 4. Subcategory
 5. Note
 6. INR
 7. Income/Expense
 8. Note.1
 9. Amount
 10. Currency
 11. Account.1

- The datatypes is number of strings is 6, and the number of floats is 5.

- The number of unique values in the dataset is:

| Column          | Number of Unique Value |
|------------------|-------|
| Date             | 240   |
| Account          | 2     |
| Category         | 13    |
| Subcategory      | 0     |
| Note             | 159   |
| INR              | 138   |
| Income/Expense   | 2     |
| Note.1          | 0     |
| Amount           | 137   |
| Currency         | 2     |
| Account.1       | 137   |


